src/common/commonutils/CommandUtils.c

```diff
@@ -29,270 +31,200 @@ static int NormalizeStatus(int status)
-    {
-        OsConfigLogDebug(log, "Cannot run command '%s', command too long (%u), ARG_MAX: %u", command, (unsigned)commandLineLength, (unsigned)maximumCommandLine);
-        return E2BIG;
-    }
-
-    commandLine = (char*)malloc(commandLineLength);
-    if (NULL == commandLine)
-    {
-        OsConfigLogError(log, "Cannot run command '%s', cannot allocate %u bytes for command, out of memory", command, (unsigned)commandLineLength);
-        return ENOMEM;
-    }
-
-    snprintf(commandLine, commandLineLength, wrappedCommand ? "%s%s%s%s" : "(%s)%s%s%s", command, commandSeparator, commandTextResultFile, commandTerminator);
-
-    // Execute the command with the requested timeout: error ETIME (62) means the command timed out
-    status = SystemCommand(context, commandLine, timeoutSeconds, callback, log);
+            if (((maxTextResultBytes > 0) && (outputBufferPos == maxTextResultBytes)) || textResult == NULL)
+            {
+                // We don't want any more data, loop to read the rest of the output.
+                continue;
+            }
 
-    free(commandLine);
+            outputBufferSize = bytesRead + outputBufferPos;
+            if ((maxTextResultBytes > 0) && (outputBufferSize > maxTextResultBytes - 1))
+            {
+                outputBufferSize = maxTextResultBytes - 1;
+            }
 
-    // Read the text result from the output of the command, if any, whether command succeeded or failed
-    if (NULL != textResult)
-    {
-        *textResult = NULL;
-        resultsFile = fopen(commandTextResultFile, "r");
-        if (resultsFile)
-        {
-            fseek(resultsFile, 0, SEEK_END);
-            fileSize = ftell(resultsFile);
-            fseek(resultsFile, 0, SEEK_SET);
+            tmp = realloc(*textResult, outputBufferSize + 1);
+            if (NULL == tmp)
+            {
+                OsConfigLogError(log, "Cannot allocate buffer for command '%s'", command);
+                status = ENOMEM;
+                free(*textResult);
+                *textResult = NULL;
+                break;
+            }
+            *textResult = tmp;
 
-            if (fileSize > 0)
+            for (inputBufferPos = 0; (inputBufferPos < bytesRead) && (outputBufferPos < outputBufferSize); inputBufferPos++, outputBufferPos++)
             {
-                // Truncate to desired maximum, if any
-                if ((maxTextResultBytes > 0) && (((size_t)fileSize + 1) > maxTextResultBytes))
+                // Copy the data. Following characters are replaced with spaces:
+                // all special characters from 0x00 to 0x1F except 0x0A (LF) when replaceEol is false
+                // plus 0x22 (") and 0x5C (\) characters that break the JSON envelope when forJson is true
+                const char c = buffer[inputBufferPos];
+                if ((replaceEol && (EOL == c)) || ((c < 0x20) && (EOL != c)) || (0x7F == c) ||
+                    (forJson && (('"' == c) || ('\\' == c))))
```

Comment:
nit: this fits on one single line, why break in two?

Reply:
Done.